project/02-Modularity/Handout.md

```diff
@@ -0,0 +1,160 @@
+# Android Project - Part 2: Modularity
+
+In this second part of the SwEng project, you will add weather forecast capabilities to your app. Using an external service, [OpenWeatherMap](https://api.openweathermap.org/), you will display to the user a weather forecast at their location, or at an address of their choice. For that, you will also need to be able to find out where the user is located.
+
+> :information_source: In this part, we will use features that were introduced in Java 8. These features were not available on older versions of Android (Android < N - API Level 24). To use them, we either need to provide alternatives, or to loose compatibility with these older devices. Here, we choose to do the latter. To enable this, please go to your app `build.gradle` and set `minSdkVersion` to `24`.
+
+Today, we will mostly focus on writing good abstractions and working modules. Handling errors won't be part of this project, so it doesn't matter if your app crashes when something goes wrong. We will learn later how to handle errors gracefully.
+
+> :information_source: **This part of the project is rather long**. Next part will be way shorter, so you can spend about 1.5 weeks on this part.
+
+## Register to OpenWeatherMap
+
+[OpenWeatherMap](https://api.openweathermap.org/) offers an API that you can access over HTTPS. As this API offers both free and paid plans, they need to be able to identify the API caller, i.e. the App that is making the HTTPS requests. This is done using a secret key, that you provide in your requests to the service, and that identifies you uniquely.
+
+To get your secret key, you will need to register to [OpenWeatherMap](https://api.openweathermap.org/). Please have a look at their [Get Started](https://openweathermap.org/appid) page to do that.
+
+## Write down the structure of the system
+
+Before diving into the code, we ask you to first write down what the system you are designing will look like. 
+
+Please take a moment to think about the modules and interfaces you will need and what kind of methods they should expose to their client (the rest of your app). We advise you to write that down on a sheet of paper, or try to explain it to a rubber duck on your desk, to convince you that it works and will fulfill its purpose.
+
+Remember, what you will code will need to be able to retrieve the location of the user, retrieve the weather at a given location, as well as convert an address to a location.
+
+## Design and write the interfaces
+
+> :warning: We advise you not to read this section before thinking about the previous one, as we will "spoil" one of the possible structure of this system.
+
+Now that you know how your system should work, you can write interfaces and data classes that _describe_ this behaviour. Don't think too much about the actual implementation, just write an interface that exposes the behaviour you need.
```

Comment:
Why "data classes" and not simply "classes" ?
Also the second sentence suggests that they should only write interfaces and not classes ?

Maybe replace the second sentence by:

Don't think too much about the actual implementation, just write a skeleton that represents the behaviour you need.